### PR TITLE
Ingestion Speedup Multiple strategy

### PR DESCRIPTION
**Pull request by lopagela**
Multiple strategy for file ingestion.

This is good to be pushed as is.
Following PR will come to add a switch between the implementation using the configuration.
Documentation should be added as well.
Ideally, the most optimal implementation should be debugged as well in this future PR.

---

The current behavior is preserved and will not impact the existing users

### CODE REVIEW

File: private_gpt/components/ingest/ingest_component.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        for file_name, file_data in files:
+            documents = IngestionHelper.transform_file_into_documents(
+                file_name, file_data
+            )
+            saved_documents.extend(self._save_docs(documents))
+        return saved_documents
+
+    def _save_docs(self, documents: list[Document]) -> list[Document]:
+        logger.debug("Modifying metadata of count=%s documents", len(documents))
+        for document in documents:
```
Comment (imartinez) [inline private_gpt/components/ingest/ingest_component.py:134]: you could use  `_exclude_metadata()` here

File: private_gpt/components/ingest/ingest_component.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+        logger.info(
+            "Transformed count=%s files into count=%s documents",
+            len(files),
+            len(documents),
+        )
+        return self._save_docs(documents)
+
+    def _save_docs(self, documents: list[Document]) -> list[Document]:
+        logger.debug("Modifying metadata of count=%s documents", len(documents))
+        for document in documents:
```
Comment (imartinez) [inline private_gpt/components/ingest/ingest_component.py:198]: you could use  `_exclude_metadata()` here

File: private_gpt/components/ingest/ingest_component.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+            processes=self.BULK_INGEST_WORKER_NUM
+        ) as pool:
+            documents = list(
+                itertools.chain.from_iterable(pool.starmap(self.ingest, files))
+            )
+        return documents
+
+    def _save_docs(self, documents: list[Document]) -> list[Document]:
+        logger.debug("Modifying metadata of count=%s documents", len(documents))
+        for document in documents:
```
Comment (imartinez) [inline private_gpt/components/ingest/ingest_component.py:294]: you could use  `_exclude_metadata()` here

File: private_gpt/server/ingest/ingest_service.py [hunk 4]
```diff
@@ -1,64 +1,30 @@
...
-    from llama_index.readers.base import BaseReader
-
-# Patching the default file reader to support other file types
-FILE_READER_CLS = DEFAULT_FILE_READER_CLS.copy()
-FILE_READER_CLS.update(
-    {
-        ".json": JSONReader,
-    }
-)
+BULK_INGEST_WORKER_NUM = max((os.cpu_count() or 1) - 1, 1)
```
Comment (imartinez) [inline private_gpt/server/ingest/ingest_service.py:23]: I think this is not used

File: settings-ingest-local.yaml
Comment (imartinez) [inline settings-ingest-local.yaml:1]: Maybe we should avoid pushing this settings file. The default setup is already local.

**Review by imartinez — CHANGES_REQUESTED** [review]
Great work! Small comments, mostly some clean up.

File: private_gpt/components/ingest/ingest_component.py (hunk 1 above)
Comment (lopagela) [reply private_gpt/components/ingest/ingest_component.py:134]: You are correct, and even going further: given that these documents come from `IngestionHelper.transform_file_into_documents`, this function is already applied :ok_hand: 

Therefore, I'm removing this code :ok_hand:

File: settings-ingest-local.yaml
Comment (lopagela) [reply settings-ingest-local.yaml:1]: Yes - but this file was setting to `mock` the llm -- I've deleted this file :ok_hand:

File: private_gpt/components/ingest/ingest_component.py [hunk 5]
```diff
@@ -130,13 +130,6 @@ def bulk_ingest(self, files: list[tuple[str, Path]]) -> list[Document]:
         return saved_documents
 
     def _save_docs(self, documents: list[Document]) -> list[Document]:
-        logger.debug("Modifying metadata of count=%s documents", len(documents))
```
Comment (imartinez) [inline private_gpt/components/ingest/ingest_component.py]: Shouldn't you call _exclude_metadata?

File: private_gpt/components/ingest/ingest_component.py [hunk 6]
```diff
@@ -194,13 +187,6 @@ def bulk_ingest(self, files: list[tuple[str, Path]]) -> list[Document]:
         return self._save_docs(documents)
 
     def _save_docs(self, documents: list[Document]) -> list[Document]:
-        logger.debug("Modifying metadata of count=%s documents", len(documents))
```
Comment (imartinez) [inline private_gpt/components/ingest/ingest_component.py]: Shouldn't you call _exclude_metadata?

File: private_gpt/components/ingest/ingest_component.py [hunk 7]
```diff
@@ -290,13 +276,6 @@ def bulk_ingest(self, files: list[tuple[str, Path]]) -> list[Document]:
         return documents
 
     def _save_docs(self, documents: list[Document]) -> list[Document]:
-        logger.debug("Modifying metadata of count=%s documents", len(documents))
```
Comment (imartinez) [inline private_gpt/components/ingest/ingest_component.py]: Shouldn't you call _exclude_metadata?

**Review by imartinez — CHANGES_REQUESTED** [review]

**Review by imartinez — APPROVED** [review]

**Comment by lopagela** [comment]
Follow up PR that fixes the parallel ingestion: https://github.com/imartinez/privateGPT/pull/1336

**Pull request merged**